Adding SiWx917 Network Stack Application APIs

ragurram26 opened the pull request:
To support the SiWx917 Network Stack Application
protocol offloading, we need to add the Silicon Labs APIs,
so adding the corresponding source files and include files

ragurram26 committed:
drivers: wifi: siwx91x: SNTP NetStack Application protocol

To support the SiWx917 SNTP  Network Stack Application protocol
offloading, we need to integrate the Silicon Labs APIs,
so adding the corresponding source files and include files
in the kconfig and CMakelist

Co-authored-by: Swami Das Nampalli <swami.das@silabs.com>
Signed-off-by: Swami Das Nampalli <swami.das@silabs.com>
Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

jhedberg commented:
@ragurram26 please provide a more detailed PR & commit message description, explaining the background to this, why this is needed, etc. Also most people in a Zephyr context won't know what "TA" is, so you need to explain that as well. If you're adding new APIs (like the title implies) I'd also expect to see some sample code to demonstrate their usage and provide code coverage for CI.

jhedberg commented:
One other thing: when you have to enumerate multiple independent changes in a single commit message (like you're doing now), it's a clear indication that you should split this up into multiple commits.

jhedberg commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -139,10 +139,10 @@ static unsigned int siwx91x_on_join(sl_wifi_event_t event,
 		net_if_dormant_off(sidev->iface);
 	}
-
+#ifndef CONFIG_WIFI_SILABS_SIWX91X_NET_STACK_OFFLOAD
```
Could this use `if (!IS_ENABLED(CONFIG_WIFI_SILABS_SIWX91X_NET_STACK_OFFLOAD)) {`?

ragurram26 replied:
Updated

ragurram26 commented:
@jhedberg I have updated the description .
> I'd also expect to see some sample code to demonstrate their usage and provide code coverage for CI.
@jhedberg Should we include the application in this pull request, or is there an alternative method to ensure code coverage for continuous integration (CI)?

ragurram26 committed:
drivers: wifi: siwx91x: HTTP NetStack Application

To support the SiWx917 HTTP  Network Stack Application
protocol offloading, we need to integrate the Silicon
Labs APIs, so adding the corresponding source files
and include files in the kconfig and CMakelist

Co-authored-by: Swami Das Nampalli <swami.das@silabs.com>
Signed-off-by: Swami Das Nampalli <swami.das@silabs.com>
Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

jerome-pouiller commented:
In the commit titles, we add space after colons: `drivers: wifi: siwx91x: SNTP ...`

jerome-pouiller commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -48,6 +48,9 @@ config NET_MGMT_EVENT_QUEUE_SIZE
 	bool "SNTP"
 
+config WIFI_SILABS_SIWX91X_HTTP_CLIENT
+	bool "HTTP"
```
Ditto.

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -116,8 +116,11 @@ int siwg91x_get_nwp_config(int wifi_oper_mode, sl_wifi_device_configuration_t *g
 	if (IS_ENABLED(CONFIG_WIFI_SILABS_SIWX91X_NET_STACK_OFFLOAD)) {
 		boot_config->tcp_ip_feature_bit_map |= SL_SI91X_TCP_IP_FEAT_ICMP;
-#ifdef CONFIG_WIFI_SILABS_SIWX91X_SNTP_CLIENT		
+#ifdef CONFIG_WIFI_SILABS_SIWX91X_SNTP_CLIENT
```
Place this change in the previous commit.

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -116,8 +116,11 @@ int siwg91x_get_nwp_config(int wifi_oper_mode, sl_wifi_device_configuration_t *g
+#endif
+#ifdef CONFIG_WIFI_SILABS_SIWX91X_HTTP_CLIENT
+		boot_config->tcp_ip_feature_bit_map |= SL_SI91X_TCP_IP_FEAT_HTTP_CLIENT;
 #endif		
```
Also remove this trailing space.

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -116,8 +116,11 @@ int siwg91x_get_nwp_config(int wifi_oper_mode, sl_wifi_device_configuration_t *g
+#ifdef CONFIG_WIFI_SILABS_SIWX91X_SNTP_CLIENT
 		boot_config->tcp_ip_feature_bit_map |= SL_SI91X_TCP_IP_FEAT_SNTP_CLIENT;
+#endif
+#ifdef CONFIG_WIFI_SILABS_SIWX91X_HTTP_CLIENT
```
`IS_ENABLED()`

jerome-pouiller commented:
In your commit message, you mention `Network Stack Application`, but I think this wording does not make sense outside of your team.

ragurram26 commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -139,10 +139,10 @@ static unsigned int siwx91x_on_join(sl_wifi_event_t event,
 		net_if_dormant_off(sidev->iface);
 	}
-
+	if (IS_ENABLED(CONFIG_WIFI_SILABS_SIWX91X_NET_STACK_OFFLOAD)) {
```
Since we want to explicitly invoke the SiWx917 IP change API from the application, we are bypassing this functionality when CONFIG_WIFI_SILABS_SIWX91X_NET_STACK_OFFLOAD is enabled.

jerome-pouiller replied:
Which application?
Can you ensure zephyr/samples/net/wifi/shell/sample.net.wifi still work?

jukkar commented:
TBH I am not a big fan of providing separate networking APIs specific to one vendor in upstream zephyr sources. I do not recall other vendors having these (at least on the networking side). Is this really needed?

ragurram26 committed:
drivers: wifi: siwx91x: MQTT NetStack Application

To support the SiWx917 MQTT Network Stack Application
protocol offloading, we need to integrate the Silicon
Labs APIs, so adding the corresponding source files
and include files in the kconfig and CMakelist

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 commented:
> @ragurram26 in your simple sample apps,can you provide the metrics demonstrating the gain of this option compared to `WIFI_SILABS_SIWX91X_NET_STACK_OFFLOAD`.

we can provide this info

jukkar commented:
> maybe this PR should include some metrics to prove it make sense to merge it.

That sounds ok. I was mostly worried that we will start to see some ifdefs in generic network samples and use either the native or vendor API, fortunately that does not seem to be case here.

ragurram26 committed:
drivers: wifi: siwx91x: ICMP NetStack Application

To support the SiWx917 ICMP Network Stack Application
protocol offloading, we need to integrate the Silicon
Labs APIs, so adding the corresponding source files
and include files in the kconfig and CMakelist

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 committed:
drivers: wifi: siwx91x: Websockets NetStack Application

To support the SiWx917 Websockets Network Stack Application
protocol offloading, we need to integrate the Silicon Labs
APIs, so adding the corresponding source files and include
files in the kconfig and CMakelist

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 committed:
drivers: wifi: siwx91x: MDNS NetStack Application

To support the SiWx917 MDNS Network Stack Application
protocol offloading, we need to integrate the Silicon Labs
APIs, so adding the corresponding source files and include
files in the kconfig and CMakelist

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 committed:
drivers: wifi: siwx91x: DNS NetStack Application

To support the SiWx917 DNS Network Stack Application
protocol offloading, we need to integrate the Silicon Labs
APIs, so adding the corresponding source files and include
files in the kconfig and CMakelist

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 committed:
drivers: wifi: siwx91x: Adding TLS bitmap

To support the SiWx917 TLS Sockets , Need to
enable the TLS bitmap in the intilization

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 committed:
drivers: wifi: siwx91x: Adding Single TLS bitmap

To support the SiWx917 Single TLS Sockets,
Need to enable the Single TLS bitmap in
the intilization

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 committed:
drivers: wifi: siwx91x: Adding Three SSL bitmap

To support the SiWx917 Three SSL Sockets,
Need to enable the Three SSL bitmap in
the intilization

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 committed:
drivers: wifi: siwx91x: Adding SSL bitmap

To support the SiWx917 SSL high streaming
throughput, Need to enable the high streaming
bitmap in the intilization

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 committed:
drivers: wifi: siwx91x: Muliple TLS version

To support the SiWx917 Multiple TLS version,
Need to enable the SSL version bitmap
in the intilization

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 committed:
drivers: wifi: siwx91x: SSL 16k record

Added support for 16K SSL record sizes, which
improves performance for SSL connections
that use larger record sizes

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 committed:
drivers: wifi: siwx91x: Adding SSL cloud memory

Added support to allocate additional memory for
SSL/TLS connections, typically required for
connections to cloud servers

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 committed:
drivers: wifi: siwx91x: Adding select bitmap

Added support to Configure number of selects in SiWx917

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

ragurram26 committed:
drivers: wifi: siwx91x: Configure sockets

Added support to Configure number of sockets in SiWx917

Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

jerome-pouiller reviewed (commented):
I believe you need to provide examples and to prove the usefulness of these features before to go further.

ragurram26 commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -71,6 +71,156 @@ config WIFI_SILABS_SIWX91X_LIMIT_PACKET_BUF_PER_STA
+
+config WIFI_SILABS_SIWX91X_WEBSOCKET_CLIENT
+	bool "Offloaded implementation of Websocket client"
+	help
```
Yes websocket can be used without http

ragurram26 commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -71,6 +71,156 @@ config WIFI_SILABS_SIWX91X_LIMIT_PACKET_BUF_PER_STA
+	  about the API.
+
+config WIFI_SILABS_SIWX91X_MDNS
+	bool "Offloaded implementation of MDNS"
```
Implementation include only mdns responder will update accordingly

ragurram26 commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -71,6 +71,156 @@ config WIFI_SILABS_SIWX91X_LIMIT_PACKET_BUF_PER_STA
+	  information about the API.
+
+config WIFI_SILABS_SIWX91X_DNS
+	bool "Offloaded implementation of DNS"
```
Yes it DNS client

ragurram26 commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -71,6 +71,156 @@ config WIFI_SILABS_SIWX91X_LIMIT_PACKET_BUF_PER_STA
+	  resources than the Zephyr one.
+
+config WIFI_SILABS_SIWX91X_ENABLE_SINGLE_SSL_SOCKET
+	bool "Offloaded implementation of single SSL socket"
```
>Does it make sense to enable this option without ENABLE_TLS?
This option is depends on the ENABLE_TLS only will update the dependency .
>BTW, I am confused with usage of TLS in previous option and SSL in this option. Does Wiseconnect really do a difference?
No difference from wiseconnect perspective .will update to TLS only

ragurram26 commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -71,6 +71,156 @@ config WIFI_SILABS_SIWX91X_LIMIT_PACKET_BUF_PER_STA
+	  SiWx91x series.
+
+config WIFI_SILABS_SIWX91X_SSL_MEMORY_CLOUD
+	bool "Offloaded implementation for the SSL/TLS cloud connections"
```
Enabling this bitmap option in the SiWx917 provides additional RAM memory. To prevent failures during communication with the cloud server, it is necessary to enable this option.

ragurram26 replied:
Cloud server certificates are typically larger in size.

ragurram26 committed:
samples: boards : Wi-Fi SiWx91x sample app

This application demonstrates support for SiWx91x
ICMP and DNS protocols

Co-authored-by: Swami Das Nampalli <swami.das@silabs.com>
Signed-off-by: Swami Das Nampalli <swami.das@silabs.com>
Signed-off-by: Rahul Gurram <rahul.gurram@silabs.com>

jerome-pouiller commented on samples/boards/siwx91x/siwx91x_sample_app/prj.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_NET_DHCPV6=n
+CONFIG_MAIN_STACK_SIZE=2048
+CONFIG_WIFI_SILABS_SIWX91X_PING=y
+CONFIG_WIFI_SILABS_SIWX91X_DNS_CLIENT=y
```
Can you provide some numbers showing the added value of `CONFIG_WIFI_SILABS_SIWX91X_PING `and `CONFIG_WIFI_SILABS_SIWX91X_DNS_CLIENT` compared to the Zephyr native implementation.

jerome-pouiller reviewed (commented):
This PR is big enough. I would suggest to focus on DNS and ping and see this can be accepted.
